src/builder.rs

```diff
@@ -0,0 +1,139 @@
+//! Builder types to compose layers and services
+
+use futures::{Future, Poll};
+use std::fmt;
+use std::marker::PhantomData;
+use std::sync::Arc;
+use tower_layer::{util::Chain, Layer, LayerExt};
+use tower_service::Service;
+
+/// Configure and build a `MakeService`
+///
+/// `ServiceBuilder` collects layers and a `MakeService` transport
+/// and produces a new `MakeService` that is wrapped by the composed
+/// layer.
+#[derive(Debug)]
+pub struct ServiceBuilder<S, L, Target, Request>
+where
+    S: Service<Target>,
+    S::Response: Service<Request>,
+{
+    maker: S,
+    layer: L,
+    _pd: PhantomData<(Target, Request)>,
+}
+
+/// Composed `MakeService` produced from `ServiceBuilder`
+#[derive(Debug)]
+pub struct ServiceBuilderMaker<S, L, Target, Request> {
+    maker: S,
+    middleware: Arc<L>,
+    _pd: PhantomData<(Target, Request)>,
+}
+
+impl<S, Target, Request> ServiceBuilder<S, Identity, Target, Request>
+where
+    S: Service<Target>,
+    S::Response: Service<Request>,
+{
+    /// Create a new `ServiceBuilder` from a `MakeService`.
+    pub fn new(maker: S) -> Self {
+        ServiceBuilder {
+            maker,
+            layer: Identity::new(),
+            _pd: PhantomData,
+        }
+    }
+}
+
+impl<S, L, Target, Request> ServiceBuilder<S, L, Target, Request>
+where
+    S: Service<Target>,
+    S::Response: Service<Request>,
+    L: Layer<S::Response, Request>,
+    Target: Clone,
+{
+    /// Add a layer to the `ServiceBuilder`.
+    pub fn add<T>(self, layer: T) -> ServiceBuilder<S, Chain<L, T>, Target, Request>
```

Comment:
Something we noticed in Linkerd2 was that switching the order of our generics from `and_then` to `push` significantly helped with compile times. I don't know why, but it may be worth checking?

For reference: https://github.com/linkerd/linkerd2-proxy/blob/7e5519619d517d28754c4492887e152700c7e916/lib/stack/src/layer.rs#L22-L46

Reply:
Looking at it again, I think it was changing from `a.and_then(b).and_then(c)` to `c.push(b).push(a)`.

Reply:
If push made it easier, I would suspect this has to do how it loops through generics for coherence checks.

Reply:
I think that the really long compile times are largely a result of how Linkerd2's `Layer` can modify the target type, which requires it to be generic over both its own target and the target of the next layer in the stack. I notice that the current `tower-layer` doesn't really support this pattern, so it may not be effected by this issue...

Reply:
`tower-layer` should support this use case? The `S` generic is not bound at all, so it can be a service w/ a different `Request` generic. Or am I missing something?